### PR TITLE
Add format command to the bot

### PR DESCRIPTION
**Pull request by Twaha-Rahman**
Since this a **really** big PR, I thought it was a good idea to do PR and keep pushing to it so others can keep reviewing the code, while I finish up the remaining thing.

State of the command :

- [x] Supports add within code block (for any language) [Adding codes inside three backticks]
- [x] Supports syntax highlighting (sometimes have trouble detecting language, but overall looks promising)
- [x] Supports formatting code block with Prettier (supports only HTML, CSS, JS, JSON, Markdown)

Please review the PR and suggest me any improvement!

### CODE REVIEW

**Comment by Twaha-Rahman** [comment]
This PR will address #17 too.

**Comment by Twaha-Rahman** [comment]
The tests in Travis aren't passsing due to #28.

File: src/commands/add-formatting.js
Comment (bradtaniguchi) [inline src/commands/add-formatting.js]: This `if else` call splits the function into two distinct paths. We could use the early return:
```
if (supportedLanguage) {
  //..
  return;
}
// the code from the else side
```

File: src/commands/add-formatting.js
Comment (bradtaniguchi) [inline src/commands/add-formatting.js]: This is nasty, it should be abstracted away into a function so the anywhere we need this formatted code, we can just call the utility function.

File: src/commands/add-formatting.js
Comment (bradtaniguchi) [inline src/commands/add-formatting.js]: This callback can probably be abstracted away into another file/function. Not sure what to name the function/file yet tho.

File: src/commands/add-formatting.js
Comment (bradtaniguchi) [inline src/commands/add-formatting.js]: Instead of using an array here and then referencing each emoji via their index, we can use an enum-like approach (the keys of the "enum" can be named to whatever makes sense, or match with their actual name):
```javascript
const reactioEmojies = {
  good: '🙂',
  bad: '🙃',
  great: '😃'
};
//..
switch(reactionEmoji) {
  case 'good':
 //..
```

File: src/commands/add-formatting.js
Comment (bradtaniguchi) [inline src/commands/add-formatting.js]: So whats the differences between these two again and the benefit we get from using both rather then just 1 or the other?

File: src/utilities/confusedMessageGenerator.js
Comment (bradtaniguchi) [inline src/utilities/confusedMessageGenerator.js]: rename this file to kabab case to match with the other files and export it on the same line, again to match the other files so far.

Add comments/docs on what each of the params are, represent and their data types.

File: src/utilities/confusedMessageGenerator.js
Comment (bradtaniguchi) [inline src/utilities/confusedMessageGenerator.js]: This can be done using a `.map` call rather than a `.forEach`

File: src/utilities/formatter.js
Comment (bradtaniguchi) [inline src/utilities/formatter.js]: export this on the first line, and add comments

File: src/utilities/formatter.js
Comment (bradtaniguchi) [inline src/utilities/formatter.js:58]: What does the stripAnsi do and why is it needed? This looks like magic to me atm.

File: src/utilities/isSuppotedByPrettier.js
Comment (bradtaniguchi) [inline src/utilities/isSuppotedByPrettier.js]: export on first line, rename file and add comments

**Review by bradtaniguchi — CHANGES_REQUESTED** [review]
Reviewed the code, most of the comments are either me just being confused, and pointing out changes that can be made for code quality (and to help me from being confused hehe)

File: src/commands/add-formatting.js
Comment (Twaha-Rahman) [reply src/commands/add-formatting.js]: `hljs` is the one which supports many language and that is why it sometimes confuses simple JS/CSS/HTML codes for other languages that nobody ever heard of. (Overall `hljs` is good but it can sometimes suffer from a little accuracy problem) As for `language-classifier`, it only supports a handful of languages and quite accurate when it is inside it's comfort zone but panic when it finds languages it doesn't support.

Another thing to note here is, `hljs` gives two guesses and `language-classifier` gives only one guess. Normally, `hljs` is the one which most of the times detects correctly but it sometimes can't properly guess simple things like CSS and HTML, that's where `language-classifier` comes in.

Overall, I think a **lot** testing needs to be done.

Comment (bradtaniguchi) [reply src/commands/add-formatting.js]: I looked a little into the `language-classifier` package, and it looks like it hasn't been maintained for **years**, nor can I find any source code, only the 0.0.1 version on npm (which interestingly, isn't mentioned in the package.json)

due to this, it might make sense to focus on using **just** the hljs lib (which also hasn't been added to the package.json) as its at least maintained, with a **ton** of community support in comparison.

File: src/commands/add-formatting.js
Comment (Twaha-Rahman) [reply src/commands/add-formatting.js]: Well spotted, should've used the modified version of guard clause...

File: src/commands/add-formatting.js
Comment (Twaha-Rahman) [reply src/commands/add-formatting.js]: I always knew, I will forget about this.... I thought I will do it just before the PR.... Sorry that you had to read this horrible code...😥

File: src/commands/add-formatting.js
Comment (Twaha-Rahman) [reply src/commands/add-formatting.js]: Done..... fixed it!

File: src/utilities/confusedMessageGenerator.js
Comment (Twaha-Rahman) [reply src/utilities/confusedMessageGenerator.js]: Done renaming the file and adding JSdoc. Do you also want me to rename the function name to `kabab case`?

File: src/utilities/confusedMessageGenerator.js
Comment (Twaha-Rahman) [reply src/utilities/confusedMessageGenerator.js]: Replaced `forEach` with `map`.

File: src/utilities/formatter.js
Comment (Twaha-Rahman) [reply src/utilities/formatter.js]: Done renaming the file and adding JSdoc.

File: src/utilities/isSuppotedByPrettier.js
Comment (Twaha-Rahman) [reply src/utilities/isSuppotedByPrettier.js]: Done!

File: src/commands/add-formatting.js
Comment (Twaha-Rahman) [reply src/commands/add-formatting.js]: Done doing so!

File: src/utilities/formatter.js
Comment (Twaha-Rahman) [reply src/utilities/formatter.js:58]: When formatting code if Prettier's parser detects a syntax error, it stops parsing and shows an error by calling `console.error`. But I catch that error and send it back to Discord as message from the bot. But the problem is that Prettier _assumes_ that their error message will be displayed only in the console, so to help the devs figure out where a syntax error occurred, they colorize their error message with ANSI. This isn't a problem when it comes to console (who doesn't like colorful errors?) but if I send it to Discord, it can't display it. That's why we have to use `stripAnsi` to remove ANSI from the error message.

Shout out to @Ryuno-Ki as he helped me solve this problem as he previously ran into it where as I had no clue what went wrong!

File: src/commands/add-formatting.js
Comment (Twaha-Rahman) [reply src/commands/add-formatting.js]: Yeah....I just checked it too. I think we should scrap `language-classifier` and just use `hljs`.

As for the packages not being linked in the `package.json`, I have fixed that.

Comment (Twaha-Rahman) [reply src/commands/add-formatting.js]: Removed `language-classifier` compleletely..

File: src/commands/add-formatting.js
Comment (bradtaniguchi) [inline src/commands/add-formatting.js]: How is this command initially called? Or is it done via some emoji reaction on any message?

If this command is called on emoji reaction we could just not use this prefix, and handle this command differently then the rest of the commands.

File: src/commands/add-formatting.js
Comment (bradtaniguchi) [inline src/commands/add-formatting.js]: This looks like we are getting the **list of values** from the `reactionObjectsObj`, if such is the case, these lines can be condensed to just:
```js
const reactionEmojis = Object.values(reactionOptionsObject):
```

PS. there is also the method `Object.entries` which will return an array of `[key, value]` from the object, which could of been used if we needed the `key` and `value`, without doing a "lookup". We don't need the `key` in this case, so we can just use `Object.values` this time :)

File: src/commands/add-formatting.js
Comment (bradtaniguchi) [inline src/commands/add-formatting.js]: change to async/await to match with most of the other code in the project.

File: src/utilities/confused-message-generator.js
Comment (bradtaniguchi) [inline src/utilities/confused-message-generator.js]: Don't use a `.map` if you don't use the return value, also if your using it to push values into another array, you could just use `.map` like so:
```js
const questions = possibleLanguages.map((language, index) => `Is it \`${language}\`? (If it is \`${language}\`, then react to the unformatted message with a ${reactionEmojies[index]} emoji)`);
```

File: src/utilities/confused-message-generator.js
Comment (bradtaniguchi) [inline src/utilities/confused-message-generator.js]: Don't use `let` if you can use `const`

File: src/utilities/formatter.js
Comment (bradtaniguchi) [inline src/utilities/formatter.js]: don't use `let` when you can use `const`

File: src/utilities/formatter.js
Comment (bradtaniguchi) [reply src/utilities/formatter.js:58]: Makes sense, and Ah good to know!

Put some of what you just told me as a comment above this line so whoever looks at this code down the line knows why its there.

Comment (Ryuno-Ki) [reply src/utilities/formatter.js:58]: Correct. We want to defer to Discord for syntax highlighting. Prettier is applying some ANSI styles as well (via Chalk).
I suggested three options:
- using a library, like stripAnsi (c.f. https://github.com/prettier/prettier/issues/2956#issuecomment-333597851 )
- use `--no-color` CLI argument (c.f. https://github.com/prettier/prettier/issues/4118#issue-303283804 )
- set `FORCE_COLOR=0` environment variable (c.f. https://github.com/prettier/prettier/issues/4118#issuecomment-371401955 )

File: src/utilities/process-reply.js
Comment (bradtaniguchi) [inline src/utilities/process-reply.js]: could change to async/await

File: src/utilities/confused-message-generator.js
Comment (Ryuno-Ki) [reply src/utilities/confused-message-generator.js]: > Don't use a `.map` if you don't use the return value

In this case, `.forEach` would be more appropriate.
Alas, I second Brad.

File: src/commands/add-formatting.js
Comment (Twaha-Rahman) [reply src/commands/add-formatting.js]: Thanks for the handy tips @bradtaniguchi !

File: src/commands/add-formatting.js
Comment (Twaha-Rahman) [reply src/commands/add-formatting.js]: This command is triggered by any user's `bot`  emoji reaction to any message.

File: src/utilities/formatter.js
Comment (Twaha-Rahman) [reply src/utilities/formatter.js:58]: Done!

File: src/commands/add-formatting.js
Comment (Twaha-Rahman) [reply src/commands/add-formatting.js]: Done!

File: src/utilities/confused-message-generator.js
Comment (Twaha-Rahman) [reply src/utilities/confused-message-generator.js]: Done!

File: src/commands/add-formatting.js
Comment (Twaha-Rahman) [reply src/commands/add-formatting.js]: Changed code to use `async/await`...

File: src/utilities/process-reply.js
Comment (Twaha-Rahman) [reply src/utilities/process-reply.js]: Done!

File: src/commands/add-formatting.js
Comment (Twaha-Rahman) [reply src/commands/add-formatting.js]: The prefix doesn't really apply to this command.... infact, I don't know if we can call it a command at all since it listens for reaction rather than messages. Maybe move it to a separate directory all together?

Comment (bradtaniguchi) [reply src/commands/add-formatting.js]: Yea I'd move this file somewhere else, since its not a command.

File: src/commands/add-formatting.js
Comment (bradtaniguchi) [inline src/commands/add-formatting.js]: add `console.error`, rather get spammed then have the code suck up errors without letting us know.

File: src/index.js
Comment (bradtaniguchi) [inline src/index.js]: I don't think we the nested `try/catch`, just `console.error` the `error` return from the fetch call and send a message back to chat if possible.

If the bot can't send error messages to chat, then its fine if it explodes, as long as the system/service managing the bot notifies someone. For the official FCC version of the bot we probably wouldn't want to spam chat with error messages, but this version can. If we supress error messages via this nested try and catch **no one** will know what sort of errors are happening or why.

Assuming we host the bot on some Paas, or FCC uses some system like Docker/kubernetes, then the "host" system will manage critical errors like this, rather then the bot itself.

File: src/index.js
Comment (bradtaniguchi) [inline src/index.js]: Same as the other comment on nested try/catch blocks

File: src/utilities/process-reply.js
Comment (bradtaniguchi) [inline src/utilities/process-reply.js]: Same as the others.

File: src/commands/add-formatting.js
Comment (Twaha-Rahman) [reply src/commands/add-formatting.js]: Done...

File: src/index.js
Comment (Twaha-Rahman) [reply src/index.js]: Done...

File: src/index.js
Comment (Twaha-Rahman) [reply src/index.js]: Done...

File: src/utilities/process-reply.js
Comment (Twaha-Rahman) [reply src/utilities/process-reply.js]: Done...

**Review by bradtaniguchi — COMMENTED** [review]
So code LGTM, but on my end when testing the bot's format command using the `:robot:` emoji on the following code:
```js
console.log("Hello World!");
```

I get the following error from the discord api:
```
DiscordAPIError: Missing Permissions
    at RequestHandler.execute (/home/bradtaniguchi/Projects/discord-bot-test/node_modules/discord.js/src/rest/RequestHandler.js:170:25)
    at processTicksAndRejections (internal/process/task_queues.js:97:5) {
  method: 'delete',
  path: '/channels/709934973997678665/messages/712192004213440546/reactions',
  code: 50013,
  httpStatus: 403
}
```

The only `delete` we are performing has to deal with the emojis in the `src/index.js` file with `.removeAll()`. I guess the bot needs [specific permissions](https://discordjs.guide/popular-topics/permissions.html) for this to work.

Removing the `await reaction.message.reactions.removeAll();` line gets the feature to work (which ends up suggesting the languages `1c` and `abnf`, but that's another story that can be fixed later (#40).


There also seems to be an error returned to the client-side if they don't select one of the provided choices in the time frame provided by process reply, but this also can be fixed later.

--- 
**So whoever has a local run of this version of the bot working locally, please let me know what I have configured wrong on my bot with this current PR so we can document what we need to turn on to get the "removeAll" method working.** Otherwise, what is the impact of removing the `removeAll` emoji call?

**Comment by Twaha-Rahman** [comment]
The bot requires requires `Manage Messages` permission to work. This is due to the bot's action of removing reactions from the message (This is done to prevent the bot from reacting multiple times to the same reaction). 

You'll have to give the bot `Manage Messages` permission either explicitly from permissions panel or by inviting it into the server using a link like this-

https://discordapp.com/oauth2/authorize?client_id=YOUR-CLIENT-ID&scope=bot&permissions=8192

(P.S. : Replace `YOUR-CLIENT-ID` with the bot's client id)

For more info go [here](https://discordjs.guide/preparations/adding-your-bot-to-servers.html#bot-invite-links).

**Comment by bradtaniguchi** [comment]
> The bot requires requires `Manage Messages` permission to work. This is due to the bot's action of removing reactions from the message (This is done to prevent the bot from reacting multiple times to the same reaction).
> 
> You'll have to give the bot `Manage Messages` permission either explicitly from permissions panel or by inviting it into the server using a link like this-
> 
> https://discordapp.com/oauth2/authorize?client_id=YOUR-CLIENT-ID&scope=bot&permissions=8192
> 
> (P.S. : Replace `YOUR-CLIENT-ID` with the bot's client id)
> 
> For more info go [here](https://discordjs.guide/preparations/adding-your-bot-to-servers.html#bot-invite-links).

Yup looks to have been missing that one! I'll open another issue to update the README with those instructions so it can be part of setup instructions for this bot :)

**Comment by Twaha-Rahman** [comment]
> Yup looks to have been missing that one! I'll open another issue to update the README with those instructions so it can be part of setup instructions for this bot :)

Yeah.... the README is getting outdated.... I think I'll get around to it after this PR...

**Review by bradtaniguchi — APPROVED** [review]
LGTM 👍 🚀 :shipit:

**Pull request merged**
